zqd/space/space.go

```diff
@@ -34,59 +35,199 @@ var (
+	opMutex       sync.Mutex
+	active        int
+	deletePending bool
+
+	wg sync.WaitGroup
+	// closed to signal non-delete ops should terminate
+	cancelChan chan struct{}
 }
 
-func Open(root, name string) (*Space, error) {
-	path := filepath.Join(root, name)
-	c, err := loadConfig(path)
+func NewManager(root string) *Manager {
+	mgr := &Manager{rootPath: root}
+
+	dirs, err := ioutil.ReadDir(root)
 	if err != nil {
-		if os.IsNotExist(err) {
-			return nil, ErrSpaceNotExist
+		// XXX log somewhere?
+		return mgr
+	}
+
+	for _, dir := range dirs {
+		if !dir.IsDir() {
+			continue
 		}
-		return nil, err
+
+		path := filepath.Join(root, dir.Name())
+		config, err := loadConfig(path)
+		if err != nil {
+			// XXX log?
+			continue
+		}
+
+		space := Space{
+			path:       path,
+			conf:       config,
+			cancelChan: make(chan struct{}, 0),
+		}
+		mgr.spaces.Store(space.Name(), &space)
 	}
-	return &Space{path, c}, nil
+
+	return mgr
 }
 
-func Create(root, name, dataPath string) (*Space, error) {
-	// XXX this should be validated before reaching here.
+func (m *Manager) Create(name, dataPath string) (api.SpacePostResponse, error) {
+	_, exists := m.spaces.Load(name)
+	if exists {
+		return api.SpacePostResponse{}, ErrSpaceExists
+	}
+
 	if name == "" && dataPath == "" {
-		return nil, errors.New("must supply non-empty name or dataPath")
+		return api.SpacePostResponse{}, errors.New("must supply non-empty name or dataPath")
 	}
 	var path string
 	if name == "" {
 		var err error
-		if path, err = fs.UniqueDir(root, filepath.Base(dataPath)); err != nil {
-			return nil, err
+		if path, err = fs.UniqueDir(m.rootPath, filepath.Base(dataPath)); err != nil {
+			return api.SpacePostResponse{}, err
 		}
+		name = filepath.Base(path)
 	} else {
-		path = filepath.Join(root, name)
+		path = filepath.Join(m.rootPath, name)
 		if err := os.Mkdir(path, 0700); err != nil {
 			if os.IsExist(err) {
-				return nil, ErrSpaceExists
+				return api.SpacePostResponse{}, ErrSpaceExists
 			}
-			return nil, err
+			return api.SpacePostResponse{}, err
 		}
-	}
-	if dataPath == "" {
 		dataPath = path
 	}
+
 	c := config{DataPath: dataPath}
 	if err := c.save(path); err != nil {
 		os.RemoveAll(path)
-		return nil, err
+		return api.SpacePostResponse{}, err
 	}
-	return &Space{path, c}, nil
+
+	space := Space{
+		path:       path,
+		conf:       c,
+		cancelChan: make(chan struct{}, 0),
+	}
+	m.spaces.Store(name, &space)
+	return api.SpacePostResponse{
+		Name:    name,
+		DataDir: dataPath,
+	}, nil
+
 }
 
-func (s Space) Name() string {
+func (m *Manager) Get(name string) (*Space, error) {
+	space, exists := m.spaces.Load(name)
+	if !exists {
+		return nil, ErrSpaceNotExist
+	}
+
+	typedSpace, ok := space.(*Space)
+	if !ok {
+		return nil, errors.New("internal error")
+	}
+	return typedSpace, nil
+}
+
+func (m *Manager) Delete(name string) error {
+	space, err := m.Get(name)
+	if err != nil {
+		return err
+	}
+
+	err = space.delete()
+	m.spaces.Delete(name)
```

Comment:
I think we should only delete it from the name map if the delete succeeds. It looks like we check for `os.IsExist` in the right places in Create, but if there's some partial delete error, I'd rather a user complain that they have a space they can't delete, than the more mysterious failure that they can't create a space with a pre-existing space name.

```suggestion
	if err != nil {
	  m.spaces.Delete(name)
	}
```